Move testing into class

joseph-palmer opened the pull request:
This PR moves data set tests into a test function defined in `DataClass` which all countries inherit. If a country requires additional specific tests they can be added by defining a `test` function (with an argument `download` to specifiy if to download data or use a snapshot, should default to `FALSE`) which first calles the `DataClass` test via `super$test(download)`.

This now means the tests regarding a dataset are contained as methods of the class, so adding a new country, with tests, should not require any additional files, only a country.R file containing a country R6 class.

I don't think it makes sense to move other tests (such as thoses inspecting a class has the minimum fields and methods) into the class methods and just have them exist in the current set up (although I am happy to reconsider this if there is a good reason for doing so). 

I have kept `test_regional-datasets` but this now initalizes each country it previously checked and just runs the test function.

News.md has been updated and version incremented in DESCRIPTION.

NOTE: If merged I will add to the wiki outlining how to add custom tests.

closes #337

seabbs commented on .lintr:
```diff
@@ -1,12 +1,12 @@
 linters: with_defaults(
     line_length_linter = line_length_linter(80),
-    cyclocomp_linter = cyclocomp_linter(complexity_limit = 25L),
+    cyclocomp_linter = cyclocomp_linter(complexity_limit = 50L),
```
eek

joseph-palmer replied:
Yeah pretty much all for the test, actual is ~32. Theres nothing new though so I imagine this would have shown up if we didn't exclude the tests from linting.

seabbs commented on R/UK.R:
```diff
@@ -495,6 +495,61 @@ UK <- R6::R6Class("UK",
+      super$test(download)
+      if (self$level == "1") {
+        data_name <- "UK level 1 with 'nhsregions=TRUE'"
+        self$nhsregions <- TRUE
```
I think we only want to test the settings currently active in the class and not all the optional settings? This does make things a bit trickier though of course.

joseph-palmer replied:
What do you mean by active functions? Like only the settings for that class, such as if the level is "1" etc. ?

seabbs replied:
yes - that its true of the validation use case where someone wants to know the integrity of their data. For the does this package work use case you want to test all possible option combinations of course. Hence the added complexity.

joseph-palmer replied:
If a user is testing a class interactively they would call `$test()` which runs the tests on that class. So if they have set it up like:

```
a <- UK$new(level = "1")
a$test(snapshot_path = "path/2/some/file.rds")
```

It would run the test on the level 1 data, e.g. `test_cleaning(), test_processing(), test_return(), specific_tests()`.

Running the package level tests with testthat calls each combination (e.g. UK level 1, UK level 2), sets up the class and then calls the test method on each one.

Will that do it? or is there something I'm missing?

seabbs replied:
Mostly. I think passing snapshot for them probably doesn't make sense and as implemented this will test all possible options and not just the ones they are using. 

The other thing this is missing is the split between testing and validating. If my interest is in whether or not my data is robust then you only need a subset of tests and probably want to test your actual class without alterations rather than using a clone of the class. 

You could argue for tests as they are and another function that runs tests only on the data that is for analysis use (so  returned maybe). The use case is then a check after you have performed all the actions you are. going to with that class rather than the testing performing the actions. Is that somewhat clear?

seabbs commented on R/shared-methods.R:
```diff
@@ -509,6 +509,158 @@ DataClass <- R6::R6Class(
+    #' (FALSE). Defaults to FALSE.
+    #' @importFrom purrr walk map
+    #' @importFrom dplyr slice_tail
+    #' @importFrom testthat test_that expect_s3_class expect_true expect_error
```
drop when moving to suggests.

seabbs commented on R/shared-methods.R:
```diff
@@ -509,6 +509,158 @@ DataClass <- R6::R6Class(
+    test = function(download = FALSE) {
+      data_name <- paste0(class(self)[1], " at level ", self$level)
+      raw_path <- paste0(
+        "custom_data/", class(self)[1],
```
I think we want this path to be user configurable and default to a temporary directory to pass CRAN.

joseph-palmer replied:
I've made this user changable but currently it defaults to the custom/data directory. What would I call a tmp directory for cran? I imagine we would do this by having an on CRAN condition which would pass different location?

seabbs commented on R/shared-methods.R:
```diff
@@ -509,6 +509,158 @@ DataClass <- R6::R6Class(
+      }
+      if (download) {
+        test_that(paste0(data_name, " downloads sucessfully"), {
+          self$download()
```
So a potential issue here is that our tests are changing the state of the class. In particular if download is called when using as snapshot. This might be quite unexpected for users and have some strange side effects. There is an argument that perhaps we should be cloning the class and running all tests on that (might be tricky) but open for discussion.

seabbs replied:
One way we could think about this is to split the tests into test and validate where test is testing the infrastructure of the class (downloading etc) and validate is testing that the end data we get is valid in the ways we expect. validate would then only work after all other class methods have been called and so not change state whilst test could change state and include the validate checks.

seabbs commented on R/shared-methods.R:
```diff
@@ -509,6 +509,158 @@ DataClass <- R6::R6Class(
+      if (download) {
+        test_that(paste0(data_name, " downloads sucessfully"), {
+          self$download()
+          walk(self$data$raw, function(data) {
```
there is quite a lot of code here. some of these things might be nicely pulled out as internal functions.

seabbs commented on R/shared-methods.R:
```diff
@@ -509,6 +509,158 @@ DataClass <- R6::R6Class(
+        expect_true(ncol(self$data$processed) >= 2)
+        self$expect_processed_cols(self$data$processed, level = self$level)
+        if (!class(self)[1] %in% c("ECDC", "WHO")) {
+          local_region <- self$clone()
```
i.e maybe we should clone at the start of testing.

joseph-palmer replied:
I need to read up on cloning

seabbs replied:
fair enough it is a tricky beast. The basic issue is do we want a call to the test method to make changes to the class that is being tested. As it stands it will in some potentially quite hard to understand ways.

joseph-palmer replied:
Agreed, I will get it to clone the class and work on that

seabbs commented on R/shared-methods.R:
```diff
@@ -509,6 +509,158 @@ DataClass <- R6::R6Class(
+        expect_true(nrow(self$data$processed) > 0)
+        expect_true(ncol(self$data$processed) >= 2)
+        self$expect_processed_cols(self$data$processed, level = self$level)
+        if (!class(self)[1] %in% c("ECDC", "WHO")) {
```
also I can't remember why we did this but its use is breaking they inheritance structure as you need to add a new class here (i.e like maybe JHU). Worth thinking about if we need this.

joseph-palmer replied:
I don't really know, I will remove and see what happens

joseph-palmer replied:
Nothing apparently

seabbs commented on R/shared-methods.R:
```diff
@@ -509,6 +509,158 @@ DataClass <- R6::R6Class(
+      self$expect_columns_contain_data(data_name)
+    },
+
+    #' @description Expect data has cleaned columns. Inherited by child
```
part of me is thinking we want these as separate functions rather than methods. DataClass is getting very full! What do you think?

joseph-palmer replied:
I think a `testing_functions.R` script would be useful and make it more verbose from within testing

seabbs reviewed (changes requested):
Really nice stuff and I think it is working as intended. Review mainly focusses on how we might want it to work rather than technical details as I think there is some thinking worth doing at this stage.

joseph-palmer commented on DESCRIPTION:
```diff
@@ -88,6 +88,7 @@ Imports:
     readxl,
     rlang,
     stringr,
+    testthat (>= 3.0.0),
```
I tried it with testthat as a suggested package but got errors when running check as the test function uses `@inhertiFrom`. Whats the way to do this here? use `package::function` notation rather than using the namespace?

seabbs replied:
yes exactly

joseph-palmer commented:
Agree with all coments:

- [ ]  I will move functions to a seperate file to make the test code more verbose
- [ ] I will make running the tests user runnable and for CRAN

Once these have been done I think it will be easier to set up the cloning and configure it with test and validate.

joseph-palmer commented:
I've moved test function into a seperate R script making the `DataClass$test()` more verbose. 

For specific tests I have changed these to go in a `specific_tests()` method which gets called after the generic tests get ran. Much like `clean_common` which I think is nicer. 

I think this is now in a better state to examine further

seabbs commented on NAMESPACE:
```diff
@@ -22,13 +22,20 @@ export(SouthAfrica)
 export(reset_cache)
 export(start_using_memoise)
 export(stop_using_memoise)
+export(test_cleaning)
```
I would have thought these should be internal?

seabbs commented on R/UK.R:
```diff
@@ -495,6 +495,50 @@ UK <- R6::R6Class("UK",
+        source <- class(self_copy)[1]
+        if (!(nchar(nhs_included_path))) {
+          nhs_included_path <- paste0(
+            "custom_data/", source,
```
as in previous review this won't work interactively if this folder doesn't exist.

joseph-palmer replied:
So I had kept it in so it could run from testthat but thinking about it I could remove the defaults from here and just have testthat pass in the file path

seabbs commented on R/shared-methods.R:
```diff
@@ -509,6 +509,53 @@ DataClass <- R6::R6Class(
+    #' By default this method is empty as if any code is required it should be
+    #' defined in a child class specific `specific_tests` method.
+    #' @param self_copy R6class the object to test
+    specific_tests = function(self_copy) {
```
I am not sure that it makes sense to use this structure as opposed to the previous inheritance method. Can you clarify the advantage?

joseph-palmer replied:
Essentially I got into a knot with making a clone of the class in the parent `test` which would be called by the countries `test` (`super$test()`) which made it difficult to grab the cloned object used in the parent `test` for further tests. Doing it this way means I can just pass the cloned class to the specific tests function for their tests. Also helped with params as I don't have to redefine them all for specific tests, just params specific to that country if needed (e.g. `nhsregions` for the UK)

seabbs commented on R/shared-methods.R:
```diff
@@ -509,6 +509,53 @@ DataClass <- R6::R6Class(
+    test = function(download = FALSE, snapshot_path = "",
+                    copy = TRUE, ...) {
+      data_name <- paste0(class(self)[1], " at level ", self$level)
+      if (copy) {
```
do we ever want it to not copy?

joseph-palmer replied:
Honestly no idea, I just put it in to play safe and allow max control, I can remove these though.

seabbs commented on R/testing_functions.R:
```diff
@@ -0,0 +1,182 @@
+#' snapshot to.
+#' @importFrom purrr map walk
+#' @importFrom dplyr slice_tail
+#' @export
```
do we want these to be exported?

seabbs commented on R/testing_functions.R:
```diff
@@ -0,0 +1,182 @@
+#' snapshot file is not found, and store a snap shot at the snapshot_path. If an
+#' existing snapshot file is found then just load this data to use in future
+#' tests.
+#' @param self The R6Class country class object to perform checks
```
maybe this should be DataClass or similar if we expect people to use these.

seabbs commented on R/testing_functions.R:
```diff
@@ -0,0 +1,182 @@
+  )
+}
+
+#' Expect download runs ok
```
Might need a more formal title

seabbs commented on R/testing_functions.R:
```diff
@@ -0,0 +1,182 @@
+#' Expect clean columns
```
maybe change the file name to test-DataClass.R or similar?

seabbs commented on R/shared-methods.R:
```diff
@@ -509,6 +509,53 @@ DataClass <- R6::R6Class(
+      } else {
+        test_class <- self
+      }
+      test_download(
```
this is nice

seabbs commented on R/shared-methods.R:
```diff
@@ -509,6 +509,53 @@ DataClass <- R6::R6Class(
+        self = test_class,
+        data_name = data_name
+      )
+      self$specific_tests(self_copy = test_class, ...)
```
if we are keeping specific_tests why not use the if exists eval logic used for dispatch in clean (so specific_test doesn't have to exist if empty)

seabbs reviewed (commented):
Looks nice. A few fussy comments and qs about exporting etc. If we are keeping specific_tests I think it should be optional rather than compulsory as for the level specific cleaning functions.

seabbs reviewed (approved):
Can you give a summary of how this is working in your mind. This is quite a big PR and so needs some thought + explanation as it is hard to review. I agree it works in principle so perhaps we can merge and work on below as individual issues etc?

A few points. 

- Is data_name really needed or can it be learnt from the class.
- Is the use of snapshot saving really working as intended. I see lots of file path construction in the internal test functions that doesn't seem ideal. 
- Perhaps the test functions should actually be public to allow people to explore testing themselves (I realise going backwards and forwards on this) . 
- Extending this some of the arg names (i.e self) could do with being a bit more detailed + the docs could do with being a little more extensive on what is actually happening in these functions. 
- Specific tests functions need some more detailed docs in general. 
- We need to discuss when optional argument tests should be triggered. Personally I think this should be only those that are defined by default in the current class with an optional override that can be used for testing.

joseph-palmer commented:
I have added some extra docs and changed some stuff around. I also made an [overview document](https://github.com/epiforecasts/covidregionaldata/blob/move-testing-into-class/new_testing.md) to outline how it all works. Plan is to delete from this repo once were happy with everything and move it into the wiki. 

I'm not sure how to seperate this into testing and validation. Most of the code called from within the class test to me looks like validation (checking columns are correct types and formats and stuff isn't missing). Actual tests of the class structures and that they contain the right fields and methods I haven't changed and are in the tests directory.

seabbs commented:
Really like the overview document. It is really great. I think it could work well as a vignette (replace the code copied in with `DataClass$test` to print the code so that it doesn't get out of date)?

seabbs commented on R/ECDC.R:
```diff
@@ -119,6 +119,28 @@ ECDC <- R6::R6Class("ECDC",
       }
+    },
+
+    #' @description Run additional tests on ECDC class. Tests ECDC has required
```
Need to mention that not designed to be run on its own.

seabbs commented on R/shared-methods.R:
```diff
@@ -509,6 +513,53 @@ DataClass <- R6::R6Class(
+    #' defined in the current class instance (FALSE). Defaults to FALSE.
+    #' @param ... Additional parameters to pass to `specific_tests`
+    #' snapshot to.
+    test = function(download = FALSE, snapshot_path = "",
```
snapshot path should default to a temporary folder/file I think rather than being empty.

seabbs commented on R/shared-methods.R:
```diff
@@ -509,6 +513,53 @@ DataClass <- R6::R6Class(
+    #' `specific_tests()`.
+    #' @param download logical. To download the data (TRUE) or use a snapshot
+    #' (FALSE). Defaults to FALSE.
+    #' @param snapshot_path character_array the path to save the downloaded data
```
this could just be to a folder and then the file could be named internally (based on class and level for example?)

seabbs commented on R/shared-methods.R:
```diff
@@ -509,6 +513,53 @@ DataClass <- R6::R6Class(
+    #' (FALSE). Defaults to FALSE.
+    #' @param snapshot_path character_array the path to save the downloaded data
+    #' or read from. Must be a path to an rds file.
+    #' @param test_all logical. Run tests with all settings (TRUE) or with those
```
sorry so fussy. This is inside a function call test, does it need to be called test_all or just all?

seabbs commented on R/shared-methods.R:
```diff
@@ -509,6 +513,53 @@ DataClass <- R6::R6Class(
+    #' @param test_all logical. Run tests with all settings (TRUE) or with those
+    #' defined in the current class instance (FALSE). Defaults to FALSE.
+    #' @param ... Additional parameters to pass to `specific_tests`
+    #' snapshot to.
```
snapshot to random text left over.

seabbs commented on R/shared-methods.R:
```diff
@@ -509,6 +513,53 @@ DataClass <- R6::R6Class(
+    #' set `level = "1"` and `localise = FALSE` the tests will be ran on level 1
+    #' data which is not localised. Rather than downloading data for a test
+    #' users can provide a path to a snapshot file of data to test instead.
+    #' Tests are ran on a clone of the class. This method calls generic tests
```
*run

seabbs commented on R/shared-methods.R:
```diff
@@ -509,6 +513,53 @@ DataClass <- R6::R6Class(
+      }
+      self_copy <- self$clone()
+      test_download(
+        cntry_obj = self_copy,
```
being a pedant. it is not always a country. I think this should be DataClass_obj or similar.

seabbs commented on R/shared-methods.R:
```diff
@@ -509,6 +513,53 @@ DataClass <- R6::R6Class(
+    #' class instance runs tests with the settings in use. For example, if you
+    #' set `level = "1"` and `localise = FALSE` the tests will be ran on level 1
+    #' data which is not localised. Rather than downloading data for a test
+    #' users can provide a path to a snapshot file of data to test instead.
```
need to mention somewhere that the snapshot is only the last 1000 rows of data.

seabbs commented on R/shared-methods.R:
```diff
@@ -509,6 +513,53 @@ DataClass <- R6::R6Class(
+
+      if ("specific_tests" %in% names(self_copy)) {
+        specific <- paste0(
+          "self$specific_tests(self_copy=self_copy, download=download, ...)"
```
spaces in the code inside the paste (I know)

seabbs replied:
+ doesn't this need to be passed `test_all`?

seabbs commented on R/test-DataClass.R:
```diff
@@ -0,0 +1,188 @@
+
+#' Test download method works correctly
+#' @description Test data can be downloaded if download = TRUE, or a requested
+#' snapshot file is not found, and store a snap shot at the snapshot_path. If an
```
snapshot_path should have `` wrapping it.

seabbs commented on R/test-DataClass.R:
```diff
@@ -0,0 +1,188 @@
+}
+
+#' Test download method works correctly
+#' @description Test data can be downloaded if download = TRUE, or a requested
```
download = TRUE should be surrounded by ``

seabbs commented on R/test-DataClass.R:
```diff
@@ -0,0 +1,188 @@
+#' snapshot file is not found, and store a snap shot at the snapshot_path. If an
+#' existing snapshot file is found then load this data to use in future tests.
+#' @param cntry_obj The DataClass object (R6Class) object to perform checks on.
+#' @param download logical check to download or use a snapshot of the data
```
caps for logical etc

seabbs commented on R/test-DataClass.R:
```diff
@@ -0,0 +1,188 @@
+#' @param cntry_obj The DataClass object (R6Class) object to perform checks on.
+#' @param download logical check to download or use a snapshot of the data
+#' @param snapshot_path character_array the path to save the downloaded
+#' snapshot to.
```
add to export and make a new docs family called test + add to pgkdown site

joseph-palmer replied:
Iv'e added a family tag of tests and an export tag on the `test-DataClass` functions, how do I add this to the pkgdown site?

seabbs commented on R/test-DataClass.R:
```diff
@@ -0,0 +1,188 @@
+#' it is a data.frame, which is not empty, has at least 2 columns and calls
+#' `expect_processed_columns` to check each column types.
+#' @inheritParams test_download
+#' @param test_all logical. Run tests with all settings (TRUE) or with those
```
caps for logical etc

seabbs commented on R/UK.R:
```diff
@@ -496,6 +496,52 @@ UK <- R6::R6Class("UK",
+      }
+      if (self_copy$level == "1") {
+        self_copy$data_name <- "UK level 1 with 'nhsregions=TRUE'"
+        self_copy$nhsregions <- TRUE
```
I would have thought these bits would only be triggered if test all was true of if the nhsregions was true?

seabbs commented on R/UK.R:
```diff
@@ -496,6 +496,52 @@ UK <- R6::R6Class("UK",
+            expect_true(nrow(self_copy$data$raw$nhs) > 0)
+            expect_true(ncol(self_copy$data$raw$nhs) >= 2)
+          })
+          self_copy$data$raw$nhs <- slice_tail(self_copy$data$raw$nhs, n = 1000)
```
we only want to test on a snapshot read from disk. If testing on the full data source should not take the tail (I think)?

joseph-palmer replied:
So add another parameter, such as `full` which if TRUE saves the full dataset to work on and if FALSE saves just the snapshots for the tests?

seabbs replied:
I was actually just thinking of some logic so that if working from the snapshot then it is partial data and if working from the raw download it is complete but yes another arg could be good but not required.

joseph-palmer replied:
Sorry, I'm confused I thought the tests always work on a snapshot of the raw data?

seabbs commented on R/test-DataClass.R:
```diff
@@ -0,0 +1,188 @@
+        testthat::expect_true(ncol(data) >= 2)
+      })
+    })
+    cntry_obj$data$raw <- map(cntry_obj$data$raw,
```
I thought we only tested on a snapshot when reading from disk. This will always trim to the last few data points which I think we don't want?

joseph-palmer replied:
I just left this as it was in the previous tests

joseph-palmer replied:
ignore me found the error with it

seabbs replied:
fair enough. Can we change here as I think this was a mistake/bug/woops.

joseph-palmer replied:
This is the same as in the tests for regular download which also uses slice_tail to get the last 250 rows

seabbs commented on new_testing.Rmd:
```diff
@@ -0,0 +1,201 @@
+---
+output: github_document
```
really like this. Keen to see as a vignette

seabbs commented on new_testing.Rmd:
````diff
@@ -0,0 +1,201 @@
+For a given country, such as the UK, you would make this by calling something like `ukdata <- UK$new(level = "1")` and can run the steps one by one by calling the respective methods: `ukdata$download(); ukdata$clean(); ukdata$process()`. To run the tests you would call `ukdata$test(shapshot_path = "path/2/data.rds")`. Rather than running tests on your active class, the code first makes a clone of ...
+
+```{r, eval = FALSE, echo = TRUE}
+test_download <- function(cntry_obj, download, snapshot_path) {
````
I'd suggest loading this code in from the package vs copying it across to make this easier to update.

seabbs commented on new_testing.md:
```diff
@@ -0,0 +1,254 @@
+
+## New testing summary
```
suggest point at this vignette from the docs for test method etc.

seabbs commented on tests/testthat/custom_tests/test-regional-dataset.R:
```diff
@@ -1,135 +1,24 @@
+    download = download,
+    snapshot_path = snapshot_path,
+    test_all = TRUE,
+    nhs_included_path = nhs_snapshot_path
```
this seems like it will get unmanageable in quite a short time. I think using a folder based path and then controlling the name internally would resolve this (as here fort example you wouldn't need a new path arg per custom data test)

seabbs replied:
there is a world where we might want this as a function in the package as well....

seabbs reviewed (changes requested):
Really nice. At the very fussy review stage but very very close. This is really slick - love it.

joseph-palmer commented:
Done the fixes marked resolved, I appreciate all the comments and I think this is looking much better. The only things I have not done are the comments left unresolved: adding to pkgdown site (unsure how to do this) and saving the full data or snapshot for tests (just a point of clarification)

seabbs commented on R/ECDC.R:
```diff
@@ -119,6 +119,29 @@ ECDC <- R6::R6Class("ECDC",
+
+    #' @description Run additional tests on ECDC class. Tests ECDC has required
+    #' additional columns and that there is only one row per country. Designed
+    #' to be ran from `test` and not ran directly.
```
ran -> run?

joseph-palmer replied:
slipped through the net

seabbs commented on R/UK.R:
```diff
@@ -496,6 +496,56 @@ UK <- R6::R6Class("UK",
+          }
+          if (download) {
+            test_that(paste(self_copy$data_name, " downloads sucessfully"), { # nolint
+              self_copy$download()
```
I know this is a straight copy but shouldn't this just call the nhs regions download code as this is actually testing all files can be downloaded (I think it is just an if else but if more complex this way of doing things makes total sense).

seabbs commented on R/shared-methods.R:
```diff
@@ -509,6 +513,65 @@ DataClass <- R6::R6Class(
+    #' for all country class objects. It also calls country specific tests
+    #' which can be defined in an individual country class method called
+    #' `specific_tests()`. The snapshots contain the first 1000 rows of data.
+    #' For more details see the 'testing' vignette.
```
mention that: can build the vignette using vignette(vignette name) + link to our hosted site version.

seabbs commented on R/shared-methods.R:
```diff
@@ -509,6 +513,65 @@ DataClass <- R6::R6Class(
+    #' (FALSE). Defaults to FALSE.
+    #' @param snapshot_dir character_array the name of a directory to save the
+    #' downloaded data or read from. If not defined a directory called
+    #' 'snapshots' will be created in the current working directory. Snapshots
```
we can't default to saving in the current directory as CRAN will get mad. We need to have the default be tempdir().

seabbs commented on R/test-DataClass.R:
```diff
@@ -0,0 +1,204 @@
+#' @description Test data can be downloaded if `download = TRUE`, or a requested
+#' snapshot file is not found, and store a snap shot in the `snapshot_dir`. If
+#' an existing snapshot file is found then load this data to use in future tests
+#' @param DataClass_obj The R6Class object to perform checks on.
```
mention and link to the fact it needs to be a `DataClass` or child.

seabbs reviewed (changes requested):
Phew lots of changes - you are right about the squash merge. This is looking really nice and I think this is the final review (yay). 

Wrapping up issues: 

- [x] A few notes on documentation
- [x] CRAN really really does not want package devs to default to saving to the users disk. The default for snapshot_dir needs to be in the temp directory by default (it might be nice to have some verbose messaging saying where the snapshot is that can be turned off). 
- [ ] For the family can set a better descriptor here: https://github.com/epiforecasts/covidregionaldata/blob/master/man/roxygen/meta.R
- [ ] For the package down site basically just need to clone one of the other sub categories (make sure to clearly say requires test that to be installed) and link using the concept roxygen tag (as for utilities etc). Ahead of utilities seems to make sense (and is a good one to copy): https://github.com/epiforecasts/covidregionaldata/blob/7e3aacbffd19846ee4185aa934b295c07cdc7859/_pkgdown.yml#L60 you can check the site locally using `pkgdown::build_site()` and looking at the docs folder (starting at index.html obviously). 

This is a fairly major PR. Do you think it would be helpful to have some eyes on this other than mine (I think no but I am very impressed with myself generally).

joseph-palmer commented:
Yeah I think it's worth casting some more eyes over this as its a big PR and change, especially with the review under way from JOSS, last thing I want to do is merge this in only for them to find bugs!

seabbs commented on R/test-DataClass.R:
```diff
@@ -0,0 +1,206 @@
+#' column is a character in the cleaned data (data$clean)
+#' @param data The clean data to check
+#' @param level character_array the level of the data to check
+#' @family tests
```
Needs a concept tag for pkgdown?

seabbs commented on vignettes/testing.Rmd:
```diff
@@ -0,0 +1,76 @@
+---
+title: "Testing"
+output: rmarkdown::html_vignette
```
Please link from docs using a url for ease

seabbs reviewed (commented):
Not sure will show up in pkgdown without concept tags.

seabbs reviewed (approved):
I think this is good for merging now. Happy for that to happen or to wait for another reviewer to take a look.
